text/0000-inline-semantic.md

```diff
@@ -0,0 +1,993 @@
+- Feature Name: `inline_semantic`, `caller_location`
+- Start Date: 2017-07-31
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+----
+
+# Summary
+[summary]: #summary
+
+Enable accurate caller location reporting during panic in `{Option, Result}::{unwrap, expect}` with
+the following changes:
+
+1. Support the `#[inline(semantic)]` function attribute, which guarantees a function is inlined
+    before reaching LLVM.
+2. Adds lang-item consts which retrieves the caller's source location.
+
+Example:
+
+```rust
+#![feature(inline_semantic, caller_location)]
+use core::caller;
+
+#[inline(semantic)]
+fn unwrap(self) -> T {
+    panic!("{}:{}:{}: oh no", caller::FILE, caller::LINE, caller::COLUMN);
+}
+
+let n: Option<u32> = None;
+let m = n.unwrap();
+```
+
+<!-- TOC -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+- [Guide-level explanation](#guide-level-explanation)
+    - [Let's reimplement `unwrap()`](#lets-reimplement-unwrap)
+    - [Semantic-inlining](#semantic-inlining)
+    - [Caller location](#caller-location)
+    - [Why do we use semantic-inlining](#why-do-we-use-semantic-inlining)
+- [Reference-level explanation](#reference-level-explanation)
+    - [Survey of panicking standard functions](#survey-of-panicking-standard-functions)
+    - [Semantic-inlining MIR pass](#semantic-inlining-mir-pass)
+    - [Caller location lang-item](#caller-location-lang-item)
+    - [Source-location-forwarding methods](#source-location-forwarding-methods)
+    - [Runtime-free backtrace for `?` operator](#runtime-free-backtrace-for--operator)
+- [Drawbacks](#drawbacks)
+    - [Code bloat](#code-bloat)
+    - [Does not support dynamic call](#does-not-support-dynamic-call)
+    - [Narrow solution scope](#narrow-solution-scope)
+- [Rationale and alternatives](#rationale-and-alternatives)
+    - [Rationale](#rationale)
+    - [Alternatives](#alternatives)
+        - [🚲 Name of everything 🚲](#🚲-name-of-everything-🚲)
```

Comment:
@ariasuni Thanks. This is a bug in the Markdown TOC generator (AlanWalk/markdown-toc#31).